eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -599,14 +599,18 @@ instance Show ImpException where
       <> displayException e
 instance Exception ImpException
 
--- | Annotation for when failure happens
+-- | Annotation for when failure happens. All the logging done within annotation will be
+-- discarded if there no failures within the annotation.
 impAnn :: NFData a => String -> ImpTestM era a -> ImpTestM era a
 impAnn msg m = do
-  catchAnyDeep m $ \exc ->
+  logs <- use impLogL
+  res <- catchAnyDeep m $ \exc ->
     throwIO $
       case fromException exc of
         Just (ImpException ann origExc) -> ImpException (msg : ann) origExc
         Nothing -> ImpException [msg] exc
+  impLogL .= logs
```

Comment:
But these scopes couldn't be related to each other?  So if we have a failure, there is no chance that the logs in this other scope could help track down the failure?

Reply:
I think the idea is that each scope is hand-crafted to help debug a specific problem, so if there is some information you would like to have in multiple scopes, you'll just turn those loggers into a function and run it again in each scope where that information is relevant.